qml/pages/ChatPage.qml

```diff
@@ -266,6 +266,17 @@ Page {
         forwardMessagesTimer.messageIds = messageIds;
         forwardMessagesTimer.start();
     }
+    function hasSendPrivilege(privilege) {
+        return chatPage.isPrivateChat ||
+                chatGroupInformation &&
+                (
+                    (chatGroupInformation.status["@type"] === "chatMemberStatusMember" && chatGroupInformation.status.permissions[privilege])
+                    || chatGroupInformation.status["@type"] === "chatMemberStatusAdministrator"
+                    || chatGroupInformation.status["@type"] === "chatMemberStatusCreator"
+                    || (chatGroupInformation.status["@type"] === "chatMemberStatusRestricted" && chatGroupInformation.status.permissions[privilege])
```

Comment:
Each `chatGroupInformation.status["@type"]` is a lookup in QVariantMap (too lookups actually, `chatGroupInformation.value("status").toMap().value("@type")`), each returning the same result. And being done from QML, it's significantly slower than its C++ equivalent. Better to calculate it once and then compare against constants. Similarly, `chatGroupInformation.status.permissions[privilege]` is 3 lookups, done twice. Even better would be to expose corresponding properties from the model and compare them against enums, at least the `@type` thing...

Reply:
Again, it's easy to forget we're not dealing with array lookups inside a current JS engine (which usually are amazingly fast depending on scope lookups) – thanks for spelling it out for me!

Reply:
done